In [ ]:
## Lab 2

# View the Data

In [1]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv('s3://ia340-2025-fall-mt-4427/BLM_Natl_MLRS_Mining_Claims_-_Not_Closed.csv') # load the data from s3
df.head() #show the top 5 rows

OBJECTID                  ID CSE_NAME      STAGE_ID    BLM_PROD  \
0         1  803391412212129792   WR 195  8.033913e+17  Lode Claim   
1         2  803392270358429696   WR 196  8.033922e+17  Lode Claim   
2         3  803393529421299712   WR 198  8.033934e+17  Lode Claim   
3         4  803394369225912320   WR 199  8.033942e+17  Lode Claim   
4         5  803394873595121664   WR 200  8.033947e+17  Lode Claim   

   CSE_TYPE_NR       CSE_NR LEG_CSE_NR               SF_ID CSE_DISP  SRC  \
0       384101  NV105216280        NaN  a02t0000006LcZyAAK   Active  LLD   
1       384101  NV105216281        NaN  a02t0000006LcZzAAK   Active  LLD   
2       384101  NV105216283        NaN  a02t0000006Lca1AAC   Active  LLD   
3       384101  NV105216284        NaN  a02t0000006Lca2AAC   Active  LLD   
4       384101  NV105216457        NaN  a02t0000006Lca3AAC   Active  LLD   

                                                QLTY  \
0  8: \n18 sections retrieved\n240 intersected fe...   
1  0: \n18 sections retrieved\n240 intersected fe...   
2  0: \n18 sections retrieved\n240 intersected fe...   
3  8: \n18 sections retrieved\n240 intersected fe...   
4  0: \n18 sections retrieved\n240 intersected fe...   

                                            CSE_META  RCRD_ACRS  \
0  NV 21 0450N 0700E 032 A NWNE|NV 21 0450N 0700E...      20.66   
1  NV 21 0450N 0700E 032 A NWNE|NV 21 0450N 0700E...      20.66   
2  NV 21 0450N 0700E 032 A SESE|NV 21 0450N 0700E...      20.66   
3  NV 21 0450N 0700E 032 A SESE|NV 21 0450N 0700E...      20.66   
4  NV 21 0450N 0700E 032 A SESE|NV 21 0450N 0700E...      20.66   

                  Created                Modified  Shape_Length  Shape_Area  
0  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.062666    0.000175  
1  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.033846    0.000070  
2  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.048314    0.000140  
3  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.062669    0.000245  
4  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.033843    0.000070

In [4]:
df.tail() #shows the last 5 rows

OBJECTID                   ID       CSE_NAME      STAGE_ID  \
565490    566765  1407167291136876544  Gold Spring 1  1.407153e+18   
565491    566766  1407052741494382592         SF 117  1.407053e+18   
565492    566767  1407076332151709696         SF 120  1.407076e+18   
565493    566768  1407090696472698880         SF 201  1.407091e+18   
565494    566769  1407097754592944128       JUJU 124  1.407098e+18   

            BLM_PROD  CSE_TYPE_NR       CSE_NR LEG_CSE_NR               SF_ID  \
565490  PLACER CLAIM       384201  MT106745768        NaN  a02SJ00000BPsLRYA1   
565491    LODE CLAIM       384101  UT106745137        NaN  a02SJ00000BGmnCYAT   
565492    LODE CLAIM       384101  UT106745140        NaN  a02SJ00000BGmnFYAT   
565493    LODE CLAIM       384101  UT106745145        NaN  a02SJ00000BGmnKYAT   
565494    LODE CLAIM       384101  NV106741757        NaN  a02SJ00000AT0BSYA1   

       CSE_DISP      SRC                                               QLTY  \
565490    Filed  PLOTTED                                                100   
565491    Filed      LLD  1; MIDPOINT, TYPICAL QUARTER SECTION WITH 4 ST...   
565492    Filed      LLD  1; MIDPOINT, TYPICAL QUARTER SECTION WITH 4 ST...   
565493    Filed      LLD  1; MIDPOINT, TYPICAL QUARTER SECTION WITH 4 ST...   
565494    Filed      LLD  1; MIDPOINT, TYPICAL QUARTER SECTION WITH 4 ST...   

                                                 CSE_META  RCRD_ACRS  \
565490  MT 20 0100N 0030E 006 A NESE | MT 20 0100N 003...  14.753654   
565491  UT 26 0420S 0180W 035 A SENW | UT 26 0420S 018...  20.660000   
565492  UT 26 0420S 0180W 035 A SENE | UT 26 0420S 018...  20.660000   
565493  UT 26 0420S 0180W 022 A NESW | UT 26 0420S 018...  20.660000   
565494  NV 21 0340N 0390E 022 A NWNE | NV 21 0340N 039...  20.660000   

                       Created                Modified  Shape_Length  \
565490  2025/08/19 00:00:00+00  2025/08/19 00:00:00+00      0.009625   
565491  2025/08/18 00:00:00+00  2025/08/18 00:00:00+00      0.032615   
565492  2025/08/18 00:00:00+00  2025/08/18 00:00:00+00      0.032615   
565493  2025/08/18 00:00:00+00  2025/08/18 00:00:00+00      0.032619   
565494  2025/08/18 00:00:00+00  2025/08/18 00:00:00+00      0.047973   

        Shape_Area  
565490    0.000003  
565491    0.000066  
565492    0.000066  
565493    0.000066  
565494    0.000137

In [2]:
df.shape #show the number of rows and columns 

(565495, 18)

This says that there are 565495 rows and 18 columns of this dataset. 

In [8]:
df['OBJECTID'] #select a column 

0              1
1              2
2              3
3              4
4              5
           ...  
565490    566765
565491    566766
565492    566767
565493    566768
565494    566769
Name: OBJECTID, Length: 565495, dtype: int64

In [3]:
df.info() # shows the data types and number of null values. 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 565495 entries, 0 to 565494
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   OBJECTID      565495 non-null  int64  
 1   ID            565495 non-null  int64  
 2   CSE_NAME      565467 non-null  object 
 3   STAGE_ID      201590 non-null  float64
 4   BLM_PROD      565495 non-null  object 
 5   CSE_TYPE_NR   565495 non-null  int64  
 6   CSE_NR        565495 non-null  object 
 7   LEG_CSE_NR    350074 non-null  object 
 8   SF_ID         565495 non-null  object 
 9   CSE_DISP      565495 non-null  object 
 10  SRC           563622 non-null  object 
 11  QLTY          563425 non-null  object 
 12  CSE_META      563106 non-null  object 
 13  RCRD_ACRS     565430 non-null  float64
 14  Created       565495 non-null  object 
 15  Modified      565495 non-null  object 
 16  Shape_Length  565495 non-null  float64
 17  Shape_Area    565495 non-null  float64
dtypes: f

18 columns exist, with some being numeric (ex, RCRD_ACRS) and some being text (ex, BLM_PROD)
First impression: this is a huge dataset of field data, combining both numeric data, ID numbers, and text names for the recorded information

In [5]:
df.describe() #statistical summary of the data

OBJECTID            ID      STAGE_ID   CSE_TYPE_NR     RCRD_ACRS  \
count  565495.000000  5.654950e+05  2.015900e+05  565495.00000  5.654300e+05   
mean   283489.541886  9.260188e+17  1.118442e+18  384123.84237  5.734941e+02   
std    163587.806174  1.877712e+17  1.672189e+17      58.54047  2.437711e+05   
min         1.000000  7.991267e+17  8.032863e+17  384101.00000 -1.000000e+00   
25%    141701.500000  7.991267e+17  9.743629e+17  384101.00000  2.066000e+01   
50%    283560.000000  7.991268e+17  1.126215e+18  384101.00000  2.066000e+01   
75%    425131.500000  1.070778e+18  1.253771e+18  384101.00000  2.066000e+01   
max    566769.000000  1.407167e+18  1.407153e+18  384401.00000  1.830487e+08   

        Shape_Length     Shape_Area  
count  565495.000000  565495.000000  
mean        0.043204       0.000119  
std         0.018736       0.000118  
min         0.000000       0.000000  
25%         0.032998       0.000067  
50%         0.034994       0.000074  
75%         0.052186       0.000139  
max         0.514983       0.010160

# Clean the Data

In [15]:
#rename column2
df = df.rename(columns={"OBJECTID": "Object_ID", "QLTY": "Quality"}) 
df.head()

Object_ID                  ID CSE_NAME      STAGE_ID    BLM_PROD  \
0          1  803391412212129792   WR 195  8.033913e+17  Lode Claim   
1          2  803392270358429696   WR 196  8.033922e+17  Lode Claim   
2          3  803393529421299712   WR 198  8.033934e+17  Lode Claim   
3          4  803394369225912320   WR 199  8.033942e+17  Lode Claim   
4          5  803394873595121664   WR 200  8.033947e+17  Lode Claim   

   CSE_TYPE_NR       CSE_NR LEG_CSE_NR               SF_ID CSE_DISP  SRC  \
0       384101  NV105216280        NaN  a02t0000006LcZyAAK   Active  LLD   
1       384101  NV105216281        NaN  a02t0000006LcZzAAK   Active  LLD   
2       384101  NV105216283        NaN  a02t0000006Lca1AAC   Active  LLD   
3       384101  NV105216284        NaN  a02t0000006Lca2AAC   Active  LLD   
4       384101  NV105216457        NaN  a02t0000006Lca3AAC   Active  LLD   

                                             Quality  \
0  8: \n18 sections retrieved\n240 intersected fe...   
1  0: \n18 sections retrieved\n240 intersected fe...   
2  0: \n18 sections retrieved\n240 intersected fe...   
3  8: \n18 sections retrieved\n240 intersected fe...   
4  0: \n18 sections retrieved\n240 intersected fe...   

                                            CSE_META  RCRD_ACRS  \
0  NV 21 0450N 0700E 032 A NWNE|NV 21 0450N 0700E...      20.66   
1  NV 21 0450N 0700E 032 A NWNE|NV 21 0450N 0700E...      20.66   
2  NV 21 0450N 0700E 032 A SESE|NV 21 0450N 0700E...      20.66   
3  NV 21 0450N 0700E 032 A SESE|NV 21 0450N 0700E...      20.66   
4  NV 21 0450N 0700E 032 A SESE|NV 21 0450N 0700E...      20.66   

                  Created                Modified  Shape_Length  Shape_Area  
0  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.062666    0.000175  
1  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.033846    0.000070  
2  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.048314    0.000140  
3  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.062669    0.000245  
4  2021/01/25 00:00:00+00  2024/10/07 08:57:14+00      0.033843    0.000070

I renamed the OBJECTID and QLTY columns to have more polished column titles, Object_ID and Quality respectively 

In [16]:
df.isnull().sum() #show the number of rows containing null value

Object_ID            0
ID                   0
CSE_NAME            28
STAGE_ID        363905
BLM_PROD             0
CSE_TYPE_NR          0
CSE_NR               0
LEG_CSE_NR      215421
SF_ID                0
CSE_DISP             0
SRC               1873
Quality           2070
CSE_META          2389
RCRD_ACRS           65
Created              0
Modified             0
Shape_Length         0
Shape_Area           0
dtype: int64

There are 28 missing values under CSE_NAME, 363905 under STAGE_ID, 215421 under LEG_CSE_NR, 1873 under SRC, 2070 under Quality, 2389 under CSE_META, 65 under RCRD_ACRS

In [24]:
df = df.dropna() #drop null values
df.isnull().sum()

Object_ID       0
ID              0
CSE_NAME        0
STAGE_ID        0
BLM_PROD        0
CSE_TYPE_NR     0
CSE_NR          0
LEG_CSE_NR      0
SF_ID           0
CSE_DISP        0
SRC             0
Quality         0
CSE_META        0
RCRD_ACRS       0
Created         0
Modified        0
Shape_Length    0
Shape_Area      0
dtype: int64

I just used dropna to drop all of the missing values

# Manipulate the Data

In [23]:
df['RCRD_ACRS'] #select and display 1 column

Series([], Name: RCRD_ACRS, dtype: float64)

In [21]:
df.sort_values(by='RCRD_ACRS') #sort data by a column

Empty DataFrame
Columns: [Object_ID, ID, CSE_NAME, STAGE_ID, BLM_PROD, CSE_TYPE_NR, CSE_NR, LEG_CSE_NR, SF_ID, CSE_DISP, SRC, Quality, CSE_META, RCRD_ACRS, Created, Modified, Shape_Length, Shape_Area]
Index: []

Here I sorted the data by the RCRD_ACRS column, which sorts by the total of recorded acres in the instance 

In [19]:
df[0:5] #slice by rows

Empty DataFrame
Columns: [Object_ID, ID, CSE_NAME, STAGE_ID, BLM_PROD, CSE_TYPE_NR, CSE_NR, LEG_CSE_NR, SF_ID, CSE_DISP, SRC, Quality, CSE_META, RCRD_ACRS, Created, Modified, Shape_Length, Shape_Area]
Index: []

In [20]:
df[df['RCRD_ACRS']>10] # select rows by values 

Empty DataFrame
Columns: [Object_ID, ID, CSE_NAME, STAGE_ID, BLM_PROD, CSE_TYPE_NR, CSE_NR, LEG_CSE_NR, SF_ID, CSE_DISP, SRC, Quality, CSE_META, RCRD_ACRS, Created, Modified, Shape_Length, Shape_Area]
Index: []

This says that 